r/aws_network_interface - read after create + add support for ipv6

DrFaust92 opened the pull request:
<!--- See what makes a good Pull Request at : https://github.com/terraform-providers/terraform-provider-aws/blob/master/.github/CONTRIBUTING.md#pull-requests --->

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

<!--- Thank you for keeping this note for the community --->

<!--- If your PR fully resolves and should automatically close the linked issue, use Closes. Otherwise, use Relates --->
Closes #12260
Closes #1707
Closes #8408
Relates #1060

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):
<!--
If change is not user facing, just write "NONE" in the release-note block below.
-->

```release-note
resource_aws_network_interface: add support for adding IPV6 addresses
```

Output from acceptance testing:

<!--
Replace TestAccXXX with a pattern that matches the tests affected by this PR.

For more information on the `-run` flag, see the `go test` documentation at https://tip.golang.org/cmd/go/#hdr-Testing_flags.
-->
```
$ make testacc TESTARGS='-run=TestAccAWSENI_'
--- PASS: TestAccAWSENI_ipv6 (239.85s)
--- PASS: TestAccAWSENI_tags (241.01s)
--- PASS: TestAccAWSENI_ipv6_count (301.36s)
--- PASS: TestAccAWSENI_sourceDestCheck (235.46s)
--- PASS: TestAccAWSENI_computedIPs (117.03s)
--- PASS: TestAccAWSENI_PrivateIpsCount (253.85s)
--- PASS: TestAccAWSENI_basic (122.11s)
--- PASS: TestAccAWSENI_disappears (100.06s)
--- PASS: TestAccAWSENI_updatedDescription (194.35s)
--- PASS: TestAccAWSENI_attached (214.29s)
--- PASS: TestAccAWSENI_ignoreExternalAttachment (173.77s)
```

ewbankkit commented on aws/resource_aws_network_interface.go:
```diff
@@ -108,6 +107,18 @@ func resourceAwsNetworkInterface() *schema.Resource {
 			},
 
 			"tags": tagsSchema(),
+			"ipv6_address_count": {
```
I think `ipv6_address_count` conflicts with `ipv6_addresses`.

DrFaust92 commented:
@bflad, i have seem to have removed `terraform-providers/aws-provider` from the reviewers. can you assist in fixing this?

ewbankkit commented:
Verified acceptance tests:

```console
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSENI_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 4 -run=TestAccAWSENI_ -timeout 120m
=== RUN   TestAccAWSENI_basic
=== PAUSE TestAccAWSENI_basic
=== RUN   TestAccAWSENI_disappears
=== PAUSE TestAccAWSENI_disappears
=== RUN   TestAccAWSENI_updatedDescription
=== PAUSE TestAccAWSENI_updatedDescription
=== RUN   TestAccAWSENI_attached
=== PAUSE TestAccAWSENI_attached
=== RUN   TestAccAWSENI_ignoreExternalAttachment
=== PAUSE TestAccAWSENI_ignoreExternalAttachment
=== RUN   TestAccAWSENI_sourceDestCheck
=== PAUSE TestAccAWSENI_sourceDestCheck
=== RUN   TestAccAWSENI_computedIPs
=== PAUSE TestAccAWSENI_computedIPs
=== RUN   TestAccAWSENI_PrivateIpsCount
=== PAUSE TestAccAWSENI_PrivateIpsCount
=== CONT  TestAccAWSENI_basic
=== CONT  TestAccAWSENI_sourceDestCheck
=== CONT  TestAccAWSENI_PrivateIpsCount
=== CONT  TestAccAWSENI_computedIPs
--- PASS: TestAccAWSENI_computedIPs (42.16s)
=== CONT  TestAccAWSENI_ignoreExternalAttachment
--- PASS: TestAccAWSENI_sourceDestCheck (42.68s)
=== CONT  TestAccAWSENI_attached
--- PASS: TestAccAWSENI_basic (45.69s)
=== CONT  TestAccAWSENI_updatedDescription
--- PASS: TestAccAWSENI_updatedDescription (71.77s)
=== CONT  TestAccAWSENI_disappears
--- PASS: TestAccAWSENI_PrivateIpsCount (119.91s)
--- PASS: TestAccAWSENI_ignoreExternalAttachment (112.27s)
--- PASS: TestAccAWSENI_disappears (40.67s)
--- PASS: TestAccAWSENI_attached (237.41s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	280.148s
```

euank commented:
I'm not a valid reviewer in this repo, but for what it's worth, this change LGTM, both code-wise and from using it. I've patched this into the plugin locally, and it's worked well for managing ipv6 addresses on my ENIs.

DrFaust92 commented:
Rebased

ewbankkit commented on aws/resource_aws_network_interface_test.go:
```diff
@@ -486,248 +607,285 @@ func testAccCheckAWSENIMakeExternalAttachment(n string, conf *ec2.NetworkInterfa
-resource "aws_vpc" "foo" {
-	cidr_block = "172.16.0.0/16"
-	enable_dns_hostnames = true
+data "aws_availability_zones" "available" {
```
Change to

```hcl
data "aws_availability_zones" "available" {
  state = "available"
 
  filter {
    name   = "opt-in-status"
    values = ["opt-in-not-required"]
  }
}
```

(see #12412, #12517) else I get the error

```
--- FAIL: TestAccAWSENI_ignoreExternalAttachment (62.85s)
    testing.go:669: Step 0 error: errors during apply:
        
        Error: Error launching source instance: Unsupported: The requested configuration is currently not supported. Please check the documentation for supported configurations.
        	status code: 400, request id: b71e5cb8-fcca-44ae-85ee-8a976eb52998
        
          on /tmp/tf-test739947919/main.tf line 41:
          (source code not available)
```

ewbankkit commented on aws/resource_aws_network_interface_test.go:
```diff
@@ -486,248 +607,285 @@ func testAccCheckAWSENIMakeExternalAttachment(n string, conf *ec2.NetworkInterfa
-		Name = "terraform-testacc-network-interface"
-	}
+const testAccAWSENIConfigBase = `
+data "aws_availability_zones" "available" {
```
Change to

```hcl
data "aws_availability_zones" "available" {
  state = "available"
 
  filter {
    name   = "opt-in-status"
    values = ["opt-in-not-required"]
  }
}
```

(see #12412, #12517), else I get the error

```
--- FAIL: TestAccAWSENI_basic (62.89s)
    testing.go:669: Step 0 error: Check failed: Check 2/8 error: expected availability_zone to be us-west-2a, but was us-west-2-lax-1a
```

DrFaust92 commented:
@ewbankkit, fixed!

ewbankkit commented:
```console
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSENI_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 2 -run=TestAccAWSENI_ -timeout 120m
=== RUN   TestAccAWSENI_basic
=== PAUSE TestAccAWSENI_basic
=== RUN   TestAccAWSENI_ipv6
=== PAUSE TestAccAWSENI_ipv6
=== RUN   TestAccAWSENI_ipv6_count
=== PAUSE TestAccAWSENI_ipv6_count
=== RUN   TestAccAWSENI_disappears
=== PAUSE TestAccAWSENI_disappears
=== RUN   TestAccAWSENI_updatedDescription
=== PAUSE TestAccAWSENI_updatedDescription
=== RUN   TestAccAWSENI_attached
=== PAUSE TestAccAWSENI_attached
=== RUN   TestAccAWSENI_ignoreExternalAttachment
=== PAUSE TestAccAWSENI_ignoreExternalAttachment
=== RUN   TestAccAWSENI_sourceDestCheck
=== PAUSE TestAccAWSENI_sourceDestCheck
=== RUN   TestAccAWSENI_computedIPs
=== PAUSE TestAccAWSENI_computedIPs
=== RUN   TestAccAWSENI_PrivateIpsCount
=== PAUSE TestAccAWSENI_PrivateIpsCount
=== RUN   TestAccAWSENI_tags
=== PAUSE TestAccAWSENI_tags
=== CONT  TestAccAWSENI_basic
=== CONT  TestAccAWSENI_tags
--- PASS: TestAccAWSENI_basic (76.21s)
=== CONT  TestAccAWSENI_PrivateIpsCount
--- PASS: TestAccAWSENI_tags (127.02s)
=== CONT  TestAccAWSENI_computedIPs
--- PASS: TestAccAWSENI_computedIPs (74.49s)
=== CONT  TestAccAWSENI_sourceDestCheck
--- PASS: TestAccAWSENI_PrivateIpsCount (152.00s)
=== CONT  TestAccAWSENI_ignoreExternalAttachment
--- PASS: TestAccAWSENI_sourceDestCheck (127.38s)
=== CONT  TestAccAWSENI_attached
--- PASS: TestAccAWSENI_ignoreExternalAttachment (115.07s)
=== CONT  TestAccAWSENI_updatedDescription
--- FAIL: TestAccAWSENI_attached (28.63s)
    testing.go:684: Step 0 error: errors during apply:
        
        Error: Error launching source instance: Unsupported: The requested configuration is currently not supported. Please check the documentation for supported configurations.
        	status code: 400, request id: 05f5ab80-b6b3-4354-8874-b0d3620e7f18
        
          on /tmp/tf-test893588910/main.tf line 39:
          (source code not available)
        
        
=== CONT  TestAccAWSENI_disappears
--- PASS: TestAccAWSENI_disappears (74.96s)
=== CONT  TestAccAWSENI_ipv6_count
--- PASS: TestAccAWSENI_updatedDescription (102.77s)
=== CONT  TestAccAWSENI_ipv6
--- PASS: TestAccAWSENI_ipv6 (76.82s)
--- PASS: TestAccAWSENI_ipv6_count (151.75s)
FAIL
FAIL	github.com/terraform-providers/terraform-provider-aws/aws	584.297s
FAIL
GNUmakefile:26: recipe for target 'testacc' failed
make: *** [testacc] Error 1
```

It looks like we need to add

```hcl
data "aws_availability_zones" "available" {
  state = "available"
 
  filter {
    name   = "opt-in-status"
    values = ["opt-in-not-required"]
  }
}
```

to `TestAccAWSENI_attached`.

DrFaust92 commented:
fixed again

ewbankkit commented:
```console
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSENI_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 2 -run=TestAccAWSENI_ -timeout 120m
=== RUN   TestAccAWSENI_basic
=== PAUSE TestAccAWSENI_basic
=== RUN   TestAccAWSENI_ipv6
=== PAUSE TestAccAWSENI_ipv6
=== RUN   TestAccAWSENI_ipv6_count
=== PAUSE TestAccAWSENI_ipv6_count
=== RUN   TestAccAWSENI_disappears
=== PAUSE TestAccAWSENI_disappears
=== RUN   TestAccAWSENI_updatedDescription
=== PAUSE TestAccAWSENI_updatedDescription
=== RUN   TestAccAWSENI_attached
=== PAUSE TestAccAWSENI_attached
=== RUN   TestAccAWSENI_ignoreExternalAttachment
=== PAUSE TestAccAWSENI_ignoreExternalAttachment
=== RUN   TestAccAWSENI_sourceDestCheck
=== PAUSE TestAccAWSENI_sourceDestCheck
=== RUN   TestAccAWSENI_computedIPs
=== PAUSE TestAccAWSENI_computedIPs
=== RUN   TestAccAWSENI_PrivateIpsCount
=== PAUSE TestAccAWSENI_PrivateIpsCount
=== RUN   TestAccAWSENI_tags
=== PAUSE TestAccAWSENI_tags
=== CONT  TestAccAWSENI_basic
=== CONT  TestAccAWSENI_ignoreExternalAttachment
--- PASS: TestAccAWSENI_basic (77.45s)
=== CONT  TestAccAWSENI_tags
--- PASS: TestAccAWSENI_ignoreExternalAttachment (114.93s)
=== CONT  TestAccAWSENI_PrivateIpsCount
--- PASS: TestAccAWSENI_tags (126.79s)
=== CONT  TestAccAWSENI_computedIPs
--- PASS: TestAccAWSENI_PrivateIpsCount (151.92s)
=== CONT  TestAccAWSENI_sourceDestCheck
--- PASS: TestAccAWSENI_computedIPs (75.15s)
=== CONT  TestAccAWSENI_disappears
--- PASS: TestAccAWSENI_disappears (74.97s)
=== CONT  TestAccAWSENI_attached
--- PASS: TestAccAWSENI_sourceDestCheck (126.19s)
=== CONT  TestAccAWSENI_updatedDescription
--- PASS: TestAccAWSENI_updatedDescription (103.28s)
=== CONT  TestAccAWSENI_ipv6_count
--- PASS: TestAccAWSENI_attached (271.79s)
=== CONT  TestAccAWSENI_ipv6
--- PASS: TestAccAWSENI_ipv6_count (153.18s)
--- PASS: TestAccAWSENI_ipv6 (77.86s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	704.046s
```

DrFaust92 commented:
@ewbankkit, as always thanks a lot!

bryceml commented:
Just curious if there is something holding this up from being merged, I'd like to be able to use this.  Last comment was about 6 weeks ago.

DrFaust92 commented:
Rebased and Refactored:

```
--- PASS: TestAccAWSENI_ipv6 (239.85s)
--- PASS: TestAccAWSENI_tags (241.01s)
--- PASS: TestAccAWSENI_ipv6_count (301.36s)
--- PASS: TestAccAWSENI_sourceDestCheck (235.46s)
--- PASS: TestAccAWSENI_computedIPs (117.03s)
--- PASS: TestAccAWSENI_PrivateIpsCount (253.85s)
--- PASS: TestAccAWSENI_basic (122.11s)
--- PASS: TestAccAWSENI_disappears (100.06s)
--- PASS: TestAccAWSENI_updatedDescription (194.35s)
--- PASS: TestAccAWSENI_attached (214.29s)
--- PASS: TestAccAWSENI_ignoreExternalAttachment (173.77s)
```

bflad commented on aws/resource_aws_network_interface_sg_attachment_test.go:
```diff
@@ -56,7 +56,7 @@ func TestAccAWSNetworkInterfaceSGAttachment_disappears(t *testing.T) {
 				Check: resource.ComposeTestCheckFunc(
 					testAccCheckAWSNetworkInterfaceSGAttachmentExists(resourceName, &networkInterface),
-					testAccCheckAWSENIDisappears(&networkInterface),
+					testAccCheckResourceDisappears(testAccProvider, resourceAwsNetworkInterface(), resourceName),
```
This test has a new failure:

```
=== CONT  TestAccAWSNetworkInterfaceSGAttachment_disappears
TestAccAWSNetworkInterfaceSGAttachment_disappears: resource_aws_network_interface_sg_attachment_test.go:42: [INFO] Got non-empty plan, as expected
TestAccAWSNetworkInterfaceSGAttachment_disappears: resource_aws_network_interface_sg_attachment_test.go:42: Check 2/2 error: Error deleting ENI: InvalidNetworkInterfaceId.Malformed: Invalid id: "sg-0beb8cdaa5e27fcd2_eni-010a1844f2ddecf3d" (expecting "eni-...")
  status code: 400, request id: e0f2e0d8-d7a8-4096-8dcf-579ac880197c
--- FAIL: TestAccAWSNetworkInterfaceSGAttachment_disappears (87.79s)
```

The resource needs to align with the type:

```suggestion
					testAccCheckResourceDisappears(testAccProvider, resourceAwsNetworkInterfaceSGAttachment(), resourceName),
```

DrFaust92 replied:
--- PASS: TestAccAWSNetworkInterfaceSGAttachment_disappears (186.56s)
Thanks!

DrFaust92 commented:
Rebased + more test refactor:

```
--- PASS: TestAccAWSENI_computedIPs (114.66s)
--- PASS: TestAccAWSENI_basic (121.19s)
--- PASS: TestAccAWSENI_attached (194.78s)
--- PASS: TestAccAWSENI_PrivateIpsCount (259.21s)
--- PASS: TestAccAWSENI_ignoreExternalAttachment (182.95s)
--- PASS: TestAccAWSENI_sourceDestCheck (238.15s)
--- PASS: TestAccAWSENI_disappears (99.56s)
--- PASS: TestAccAWSENI_updatedDescription (191.10s)
--- PASS: TestAccAWSENI_ipv6_count (298.87s)
--- PASS: TestAccAWSENI_tags (238.05s)
--- PASS: TestAccAWSENI_ipv6 (248.03s)
```

bflad reviewed (approved):
Looks great, thanks @DrFaust92 🚀 

Output from acceptance testing in AWS Commercial:

```
--- PASS: TestAccAWSENI_attached (183.12s)
--- PASS: TestAccAWSENI_basic (102.76s)
--- PASS: TestAccAWSENI_computedIPs (101.72s)
--- PASS: TestAccAWSENI_disappears (87.26s)
--- PASS: TestAccAWSENI_ignoreExternalAttachment (185.44s)
--- PASS: TestAccAWSENI_ipv6 (141.80s)
--- PASS: TestAccAWSENI_ipv6_count (159.75s)
--- PASS: TestAccAWSENI_PrivateIpsCount (151.93s)
--- PASS: TestAccAWSENI_sourceDestCheck (143.44s)
--- PASS: TestAccAWSENI_tags (145.50s)
--- PASS: TestAccAWSENI_updatedDescription (123.99s)

--- PASS: TestAccAWSNetworkInterfaceSGAttachment_basic (70.59s)
--- PASS: TestAccAWSNetworkInterfaceSGAttachment_DataSource (119.79s)
--- PASS: TestAccAWSNetworkInterfaceSGAttachment_disappears (108.71s)
--- PASS: TestAccAWSNetworkInterfaceSGAttachment_Instance (139.32s)
--- PASS: TestAccAWSNetworkInterfaceSGAttachment_Multiple (90.30s)
```

Output from acceptance testing in AWS GovCloud (US):

```
--- PASS: TestAccAWSENI_attached (178.91s)
--- PASS: TestAccAWSENI_basic (95.41s)
--- PASS: TestAccAWSENI_computedIPs (93.75s)
--- PASS: TestAccAWSENI_disappears (85.16s)
--- PASS: TestAccAWSENI_ignoreExternalAttachment (146.43s)
--- PASS: TestAccAWSENI_ipv6 (131.01s)
--- PASS: TestAccAWSENI_ipv6_count (141.75s)
--- PASS: TestAccAWSENI_PrivateIpsCount (85.91s)
--- PASS: TestAccAWSENI_sourceDestCheck (131.61s)
--- PASS: TestAccAWSENI_tags (130.60s)
--- PASS: TestAccAWSENI_updatedDescription (112.40s)

--- PASS: TestAccAWSNetworkInterfaceSGAttachment_basic (68.81s)
--- PASS: TestAccAWSNetworkInterfaceSGAttachment_DataSource (86.84s)
--- PASS: TestAccAWSNetworkInterfaceSGAttachment_disappears (102.20s)
--- PASS: TestAccAWSNetworkInterfaceSGAttachment_Instance (121.17s)
--- PASS: TestAccAWSNetworkInterfaceSGAttachment_Multiple (83.78s)
```
